Simplified/NYPLBarcode.swift

```diff
@@ -0,0 +1,72 @@
+import Foundation
+import ZXingObjC
+
+fileprivate let barcodeHeight: CGFloat = 100
+fileprivate let maxBarcodeWidth: CGFloat = 500
+
+@objc enum NYPLBarcodeType: Int {
+  case codabar
+  case code39
+  case qrCode
+  case code128
+}
+
+fileprivate func ZXBarcodeFormatFor(_ NYPLBarcodeType:NYPLBarcodeType) -> ZXBarcodeFormat {
+  switch NYPLBarcodeType {
+  case .codabar:
+    return kBarcodeFormatCodabar
+  case .code39:
+    return kBarcodeFormatCode39
+  case .qrCode:
+    return kBarcodeFormatQRCode
+  case .code128:
+    return kBarcodeFormatCode128
+  }
+}
+
+/// Manage creation and scanning of barcodes on library cards.
+/// Keep any third party dependency abstracted out of the main app.
+final class NYPLBarcode: NSObject {
+
+  class func image(fromString string: String, superviewWidth: CGFloat, type: NYPLBarcodeType) -> UIImage?
+  {
+    let width = imageWidthFor(superviewWidth)
+    guard let writer = ZXMultiFormatWriter.writer() as? ZXWriter else { return nilWithGenericError() }
+    do {
+      let result = try writer.encode(string,
+                                     format: ZXBarcodeFormatFor(type),
+                                     width: Int32(width),
+                                     height: Int32(barcodeHeight))
+      if let cgImage = ZXImage.init(matrix: result).cgimage {
+        return UIImage.init(cgImage: cgImage)
+      } else {
+        return nilWithGenericError()
+      }
+    } catch {
+      Log.error(#file, "Failed to create barcode image: \(error.localizedDescription)")
+      return nil
+    }
+  }
+
+  class func presentScanner(withCompletion completion: @escaping (String?) -> ())
+  {
+    guard let scannerVC = NYPLBarcodeScanningViewController.init(completion: completion) else { return }
+    let navController = UINavigationController.init(rootViewController: scannerVC)
+    NYPLRootTabBarController.shared().safelyPresentViewController(navController, animated: true, completion: nil)
+  }
+
+  private class func imageWidthFor(_ superviewWidth: CGFloat) -> CGFloat
+  {
+    if superviewWidth > maxBarcodeWidth {
+      return maxBarcodeWidth
```

Comment:
can there be a circumstance where something goes wrong, the barcode image is generated as empty, and so there needs to also be a minWidth, so the user can see the broken image icon?

Reply:
I really have no idea if the framework would produce an "image" that does not have a barcode, but it seems unlikely. From the code I've seen the common fail-case is to return null for any number of possible issues, which I account for at the call-site. And based on the context of the feature, I'm okay completely hiding the functionality if this is the case.